test/e2e/query_test.go

```diff
@@ -849,6 +849,207 @@ func TestSidecarStorePushdown(t *testing.T) {
 	})
 }
 
+type seriesWithLabels struct {
+	intLabels labels.Labels
+	extLabels labels.Labels
+}
+
+func TestQueryStoreDedup(t *testing.T) {
+	t.Parallel()
+
+	e, err := e2e.New(e2e.WithName("storededup"))
+	testutil.Ok(t, err)
+	t.Cleanup(e2ethanos.CleanScenario(t, e))
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
+	t.Cleanup(cancel)
+
+	bucket := "store-gw-dedup-test"
+	minio := e2edb.NewMinio(e, "thanos-minio", bucket, e2edb.WithMinioTLS())
+	testutil.Ok(t, e2e.StartAndWaitReady(minio))
+
+	l := log.NewLogfmtLogger(os.Stdout)
+	bkt, err := s3.NewBucketWithConfig(l, e2ethanos.NewS3Config(bucket, minio.Endpoint("http"), minio.Dir()), "test")
+	testutil.Ok(t, err)
+
+	storeGW := e2ethanos.NewStoreGW(
+		e,
+		"s1",
+		client.BucketConfig{
+			Type:   client.S3,
+			Config: e2ethanos.NewS3Config(bucket, minio.InternalEndpoint("http"), minio.InternalDir()),
+		},
+		"",
+		nil,
+	)
+	testutil.Ok(t, e2e.StartAndWaitReady(storeGW))
+
+	tests := []struct {
+		extReplicaLabel  string
+		intReplicaLabel  string
+		desc             string
+		blockFinderLabel string
+		series           []seriesWithLabels
+	}{
+		{
+			desc:            "Deduplication works with external label",
+			extReplicaLabel: "replica",
+			series: []seriesWithLabels{
+				{
+					intLabels: labels.FromStrings("__name__", "simple_series"),
+					extLabels: labels.FromStrings("replica", "a"),
+				},
+				{
+					intLabels: labels.FromStrings("__name__", "simple_series"),
+					extLabels: labels.FromStrings("replica", "b"),
+				},
+			},
+			blockFinderLabel: "dedupext",
+		},
+		{
```

Comment:
Why this specific scenario with 4 samples, @fpetkovski? This one I have here already causes deduplication failure when using internal labels, with only 2 samples. 🤔

Reply:
FYI added below, in the "Deduplication works with extra internal label" test.

Reply:
You mentioned that that case is fixed by https://github.com/thanos-io/thanos/pull/6296, but this new one should not be.

Reply:
Ahhh, that's good info. Let me bring in those changes and try it out. 👍

Reply:
With your changes from that PR all tests pass:
<img width="470" alt="image" src="https://github.com/thanos-io/thanos/assets/159076/c5b20831-691f-4228-b716-59a81578cdea">

Reply:
Hm, interesting. Let me take a look

Reply:
Is the test setting up one block per series? Because all series need to be in the same block to reproduce the bug.

Reply:
Also the replica label should come before `my_label` when sorted. So maybe we can rename `my_label` to `z_label`.

Reply:
All tests green with `my_label` renamed to `z_label`. Still using a separate block per series.

Reply:
I'll try to hack something to share the block, but I don't think it breaks your fix. Give me some time.

Reply:
Kind of difficult to set up this scenario, @fpetkovski. There aren't helper function to set up a block in this fashion and this is starting to snowball. 

Given that I can reproduce deduplication bugs even with the series in separate blocks, WDYT we try to make these tests pass and later add this scenario where all the series are in the same block?

Reply:
Example of a failure reproducing the dedup issue: https://github.com/thanos-io/thanos/actions/runs/5024929278/jobs/9011235641

<img width="1092" alt="Screenshot 2023-05-19 at 19 08 50" src="https://github.com/thanos-io/thanos/assets/159076/3091ff10-f357-4af7-8540-56ce24827578">

Reply:
I thought you could pass all the series when creating the block instead of iterating through them: https://github.com/thanos-io/thanos/pull/6377/files#diff-e69e8c48b08313e7db23f15372816d8db13fad6da546be53e3947a00f6959145R1014.

Reply:
The helpers we have will ask for all the labels beforehand and the block will contain all of them. I cannot put few samples in a block with a labelset (let's say `a=1, b=2`) and another one with an orthogonal labelset ((let's say `c=3, d=4`).

Reply:
To do what you want, I would have to big deep into all the helpers used by `e2eutil.CreateBlockWithBlockDelay` under the hood to add this feature.

Reply:
It is valuable to have this extra case, but here we have enough tests to reproduce and fix the bug. Let's focus on getting these tests working (i.e. is your fix #6296 how we want to go about this?), so that Thanos v0.32 can be realized without this problem and people can upgrade to the `main` builds if they want to. 

In another PR we can add the case you want.

Reply:
Most users that have reported the bug remote write from pairs of Prometheus instances into a Receiver, so Prometheus external labels become Receiver internal labels in the same block. Since we don't have a test case for that scenario, I would not say that making these tests pass will fix the bug.

We can merge this PR and continue working in another branch, but https://github.com/thanos-io/thanos/pull/6296 will not solve the problem since that is not the case most people are experiencing based on comments in the issue: https://github.com/thanos-io/thanos/issues/6257#issuecomment-1494804948, https://github.com/thanos-io/thanos/issues/6257#issuecomment-1506886834, https://github.com/thanos-io/thanos/issues/6257#issuecomment-1510307684.

Reply:
@fpetkovski I also have a test using Receive that covers dedup on block label, external label, and both together: https://github.com/thanos-io/thanos/pull/6377/files#diff-61577592f26fcb88128e3bd22b45c67147d6a16f64de73aff81f79b4894a7a3aR97-R197. That's covered.

Reply:
From what I see, the problem only appears when Receive uploads the block to object storage and Store GW starts to be responsible for querying that data. While the data is "hot" in Receive the issue doesn't happen. That Receive e2e test there proves it.

Reply:
https://github.com/thanos-io/thanos/issues/6257#issuecomment-1506744862 seems to be indicating the problem happens in Receive.